examples/printerdemo_mcu/esp-idf/main/main.cpp

```diff
@@ -0,0 +1,101 @@
+// Copyright © SixtyFPS GmbH <info@slint.dev>
+// SPDX-License-Identifier: MIT
+
+#include "esp_slint.h"
+#include "printerdemo.h"
+#include <ctime>
+#include <memory>
+
+#include <slint-platform.h>
+
+#include <bsp/display.h>
+#include <bsp/esp-bsp.h>
+#include <bsp/touch.h>
+#include <vector>
+
+struct InkLevelModel : slint::Model<InkLevel>
+{
+    size_t row_count() const override { return m_data.size(); }
+    std::optional<InkLevel> row_data(size_t i) const override
+    {
+        if (i < row_count())
+            return { m_data[i] };
+        return {};
+    }
+
+    std::vector<InkLevel> m_data = { { slint::Color::from_rgb_uint8(255, 255, 0), 0.9 },
+                                     { slint::Color::from_rgb_uint8(0, 255, 255), 0.5 },
+                                     { slint::Color::from_rgb_uint8(255, 0, 255), 0.8 },
+                                     { slint::Color::from_rgb_uint8(0, 0, 0), 0.1 } };
+};
+
+extern "C" void app_main(void)
+{
+
+    /* Initialize I2C (for touch and audio) */
+    bsp_i2c_init();
+
+    /* Initialize display  */
+    esp_lcd_panel_io_handle_t io_handle = NULL;
+    esp_lcd_panel_handle_t panel_handle = NULL;
+    const bsp_display_config_t bsp_disp_cfg = {
+        .max_transfer_sz = (BSP_LCD_H_RES * CONFIG_BSP_LCD_DRAW_BUF_HEIGHT) * sizeof(uint16_t),
+    };
+    bsp_display_new(&bsp_disp_cfg, &panel_handle, &io_handle);
+    esp_lcd_touch_handle_t touch_handle = NULL;
+    const bsp_touch_config_t bsp_touch_cfg = {};
+    bsp_touch_new(&bsp_touch_cfg, &touch_handle);
+
+    /* Set display brightness to 100% */
+    bsp_display_backlight_on();
+
+    static std::vector<slint::platform::Rgb565Pixel> buffer(BSP_LCD_H_RES * BSP_LCD_V_RES);
+
+    slint::platform::set_platform(
+            std::make_unique<EspPlatform>(slint::PhysicalSize({ BSP_LCD_H_RES, BSP_LCD_V_RES }),
+                                          panel_handle, touch_handle, buffer));
+
```

Comment:
This is basically the API we expose.  the rest is initialiation that is kind of board specific.  (although everything is abstracted in `bsp`, but it depends on which bsp one loads)